### PR TITLE
plan: consider multiple in expressions when calculating range.

### PR DESCRIPTION
**Pull request by hanfei1991**
Consider the query "SELECT * FROM T WHERE a in (1,2) and b in(3,4)". Using old logic we can only get the range (1,2), But we should calculate the range as ((1,3), (1,4)), ((2,3), (2,4)).
@shenli @coocood @zimulala @XuHuaiyu @winoros PTAL

The test will pass after @coocood fix the issue #1959

### CODE REVIEW

**Comment by zimulala** [comment]
Fix CI

File: plan/physical_plans.go [hunk 1]
```diff
@@ -50,6 +50,9 @@ type PhysicalIndexScan struct {
 	// If the query requires the columns that don't belong to index, DoubleRead will be true.
 	DoubleRead bool
 
+	// All conditions in AccessCondition[accessEqualCount:accessInAndEqCount] are in expressions.
```
Comment (coocood) [inline plan/physical_plans.go]: are IN expressions.

File: plan/refiner.go [hunk 2]
```diff
@@ -248,6 +279,21 @@ func (c *conditionChecker) check(condition expression.Expression) bool {
 	return false
 }
 
+func (c *conditionChecker) extractEqOrInFunc(conditions []expression.Expression) int {
```
Comment (coocood) [inline plan/refiner.go]: rename to `findEqOrInFunc` is better.

File: plan/refiner.go [hunk 3]
```diff
@@ -44,17 +44,31 @@ func buildIndexRange(p *PhysicalIndexScan) error {
 			p.Ranges = rb.appendIndexRanges(p.Ranges, point, tp)
 		}
 	}
+	// Build rangePoints for in conditions.
+	for i := p.accessEqualCount; i < p.accessInAndEqCount; i++ {
```
Comment (shenli) [inline plan/refiner.go]: Processing for equal conditions and in conditions should keep the same style.

File: plan/refiner.go [hunk 4]
```diff
@@ -120,6 +134,9 @@ func getEQFunctionOffset(expr expression.Expression, cols []*model.IndexColumn)
 func detachIndexScanConditions(conditions []expression.Expression, indexScan *PhysicalIndexScan) ([]expression.Expression, []expression.Expression) {
 	accessConds := make([]expression.Expression, len(indexScan.Index.Columns))
 	var filterConds []expression.Expression
+	for i, cond := range conditions {
```
Comment (shenli) [inline plan/refiner.go]: Add comment for this.

File: plan/refiner.go [hunk 5]
```diff
@@ -139,36 +156,50 @@ func detachIndexScanConditions(conditions []expression.Expression, indexScan *Ph
...
+		checker := &conditionChecker{
+			tableName:    indexScan.Table.Name,
+			idx:          indexScan.Index,
+			columnOffset: curIndex,
 		}
-		cond = pushDownNot(cond, false)
-		if indexScan.accessEqualCount >= len(indexScan.Index.Columns) ||
-			!checker.check(cond) {
-			filterConds = append(filterConds, cond)
+		// First of all we should extract all of in/eq expressions. We still need to consider eq functions,
```
Comment (shenli) [inline plan/refiner.go]: The comment is not clear enough.

File: plan/refiner.go (hunk 2 above)
Comment (shenli) [inline plan/refiner.go]: Add comment for this function.

File: plan/refiner.go [hunk 6]
```diff
@@ -139,36 +156,50 @@ func detachIndexScanConditions(conditions []expression.Expression, indexScan *Ph
...
 			continue
 		}
-		accessConds = append(accessConds, cond)
-		if indexScan.Index.Columns[indexScan.accessEqualCount].Length != types.UnspecifiedLength ||
-			// TODO: it will lead to repeated compution cost.
-			checker.shouldReserve {
-			filterConds = append(filterConds, cond)
-			checker.shouldReserve = false
+		// If we fail to find any in or eq expression, we should consider all of other conditions for the next column.
+		for _, cond := range conditions {
```
Comment (coocood) [inline plan/refiner.go]: Extract this block to a function and put it in `if accessIdx == -1`, then we don't need to call `continue`.

File: plan/refiner.go [hunk 7]
```diff
@@ -139,36 +140,50 @@ func detachIndexScanConditions(conditions []expression.Expression, indexScan *Ph
...
-		columnOffset: indexScan.accessEqualCount,
-	}
-	for _, cond := range conditions {
-		isAccess := false
-		for _, acCond := range accessConds {
-			if cond == acCond {
-				isAccess = true
+	indexScan.accessInAndEqCount = indexScan.accessEqualCount
+	// We should remove all accessConds , so that they will not be added to filter conditions.
+	for i := len(conditions) - 1; i >= 0; i-- {
```
Comment (coocood) [inline plan/refiner.go]: extract this to a function

**Comment by hanfei1991** [comment]
@shenli @coocood PTAL

File: plan/physical_plans.go [hunk 8]
```diff
@@ -50,6 +50,9 @@ type PhysicalIndexScan struct {
 	// If the query requires the columns that don't belong to index, DoubleRead will be true.
 	DoubleRead bool
 
+	// All conditions in AccessCondition[accessEqualCount:accessInAndEqCount] are IN expressions.
```
Comment (shenli) [inline plan/physical_plans.go]: IN expressions or equal condition?

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**